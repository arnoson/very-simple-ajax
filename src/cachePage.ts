export const cache: Record<string, Document> = {}

export const cachePage = (url: string) =>
  (cache[url] = document.cloneNode(true) as Document)
