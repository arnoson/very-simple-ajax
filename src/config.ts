import { Config } from './types'

export const globalConfig: Required<Config> = {
  watchHistory: true,
  merge: 'replace',
}
