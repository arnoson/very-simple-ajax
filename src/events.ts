import { LinksEvent } from './types'

const handlers: Record<string, Function[]> = {}

export const on = (event: LinksEvent, handler: Function) => {
  handlers[event] ??= []
  handlers[event].push(handler)
}

export const off = (event: LinksEvent, handler: Function) => {
  handlers[event]?.splice(handlers[event].indexOf(handler), 1)
}

export const emit = async (event: LinksEvent, payload: { url: string }) => {
  for (const handler of handlers[event] ?? []) {
    await handler(payload)
  }
}
