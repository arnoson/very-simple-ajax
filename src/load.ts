const parser = new DOMParser()
let progress = 0

export const load = async (url: string): Promise<Document | undefined> => {
  try {
    setProgress(0)
    // Only show the progress bar if the page takes more than 500ms to load.
    const progressDelayTimeout = window.setTimeout(
      () => toggleLoading(true),
      500
    )

    const response = await fetch(url, { headers: { 'X-Very-Simple': '1' } })
    const reader = response.body?.getReader()
    const length = parseInt(response.headers.get('Content-Length') ?? '0')
    let receivedBytes = 0

    const stream = new ReadableStream({
      start(controller) {
        const read = async () => {
          const { done, value } = await reader!.read()

          if (done) {
            controller.close()
            return
          }

          receivedBytes += value.length
          setProgress(receivedBytes / length)
          controller.enqueue(value)

          read()
        }
        read()
      },
    })

    const html = await new Response(stream, {
      headers: { 'Content-Type': 'text/html' },
    }).text()

    clearTimeout(progressDelayTimeout)
    setProgress(1)
    toggleLoading(false)

    return parser.parseFromString(html, 'text/html')
  } catch (e) {
    // There was a network error. We reload the page so the user sees the
    // browser's network error page.
    window.location.reload()
  }
}

const setProgress = (value: number) => {
  progress = value
  document.documentElement.style.setProperty(
    '--simple-progress',
    `${value * 100}%`
  )
}

const toggleLoading = (state: boolean) =>
  document.documentElement.toggleAttribute('data-simple-loading', state)
