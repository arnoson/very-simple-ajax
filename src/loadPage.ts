const parser = new DOMParser()

export type LoadPageResponse = {
  document?: Document
  error?: Error
  response?: Response
}

/**
 * Load and parse an html page.
 * @throws
 */
export const loadPage = async (url: string): Promise<LoadPageResponse> => {
  try {
    // Todo: show progress (https://javascript.info/fetch-progress)
    const response = await fetch(url)
    const document = parser.parseFromString(await response.text(), 'text/html')
    return { document, response }
  } catch (error) {
    return { error }
  }
}
