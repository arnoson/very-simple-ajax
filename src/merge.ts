import { MergeStrategy, Regions } from './types'
// @ts-ignore (missing types)
import { Idiomorph } from 'idiomorph/dist/idiomorph.esm.js'

export const merge = (
  region: HTMLElement,
  newRegion: HTMLElement,
  strategy: MergeStrategy
): { autoFocusEl: HTMLElement | undefined } => {
  let autoFocusEl: HTMLElement | undefined

  if (strategy === 'replace') {
    region.replaceWith(newRegion)
  } else if (strategy === 'before') {
    region.insertAdjacentElement('beforebegin', newRegion)
  } else if (strategy === 'after') {
    region.insertAdjacentElement('afterend', newRegion)
  } else if (strategy === 'prepend') {
    region.insertAdjacentElement('afterbegin', newRegion)
  } else if (strategy === 'append') {
    region.insertAdjacentElement('beforeend', newRegion)
  } else if (strategy === 'update') {
    region.replaceChildren(...Array.from(newRegion.children))
  } else if (strategy === 'morph') {
    const result = morph(region, newRegion)
    autoFocusEl = result.autoFocusEl
  }

  // Find the first auto focus element, where [data-simple-autofocus] wins over
  // [autofocus].

  // First test the new region itself ...
  if (newRegion.hasAttribute('data-simple-autofocus') || newRegion.autofocus) {
    autoFocusEl ??= newRegion
  }
  // ... then look into it's children.
  autoFocusEl ??=
    newRegion.querySelector<HTMLElement>('[data-simple-autofocus]') ??
    newRegion.querySelector<HTMLElement>('[autofocus]') ??
    undefined

  return { autoFocusEl }
}

const morph = (container: Element, newContainer: Element) => {
  const manualReplacements: [HTMLElement, HTMLElement][] = []
  let autoFocusEls: HTMLElement[] = []

  Idiomorph.morph(container, newContainer, {
    callbacks: {
      afterNodeMorphed(_: Node, newNode: Node) {
        if (
          newNode instanceof HTMLElement &&
          (newNode.autofocus || newNode.dataset.simpleAutofocus)
        ) {
          autoFocusEls.push(newNode)
        }
      },
      beforeNodeMorphed(oldNode: Node, newNode: Node) {
        // All following checks use dataset, which is only available on
        // HTMLElements.
        if (!(oldNode instanceof HTMLElement)) return true
        if (!(newNode instanceof HTMLElement)) return true

        // Check if the old node is permanent and the new one is matching.
        if (
          oldNode.hasAttribute('data-simple-permanent') &&
          oldNode.id === newNode.id
        )
          return false

        // Each node could be a Very Simple Component that attached event
        // listeners to itself or its children. In this case we have to replace
        // the nodes manually instead of morphing in order to cleanup the
        // listeners.
        const oldComponent = oldNode.dataset.simpleComponent
        const newComponent = newNode.dataset.simpleComponent

        // Nothing to clean up, if the old node isn't a component are both nodes
        // are the same component.
        if (!oldComponent || oldComponent === newComponent) return true

        manualReplacements.push([oldNode, newNode])
      },
    },
  })
  for (const [oldEl, newEl] of manualReplacements) oldEl.replaceWith(newEl)

  // Find the first auto focus element, where [data-simple-autofocus] wins over
  // [autofocus].
  const autoFocusEl =
    autoFocusEls.find((el) => el.dataset.simpleAutofocus) ??
    autoFocusEls.find((el) => el.autofocus)

  return { autoFocusEl }
}
