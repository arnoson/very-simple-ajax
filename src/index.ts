import { emit, off, on } from './events'
import { cache, render } from './render'
import { Config, VisitOptions } from './types'

let previousUrl: string | undefined

export default {
  on,
  off,

  start({ watchHistory = true } = {} as Config) {
    if (watchHistory) {
      window.addEventListener('popstate', async () => {
        await this.visit(window.location.pathname, {
          action: 'none',
          useCache: true,
        })
      })
    }
  },

  async visit(
    url: string,
    { action = 'push', useCache, cacheId }: VisitOptions = {}
  ) {
    await emit('before-visit', { url })

    cacheId = cacheId ?? previousUrl ?? window.location.pathname
    cache[cacheId] = document.cloneNode(true) as Document
    await render(url, useCache)

    if (action === 'replace') {
      history.replaceState(null, '', url)
    } else if (action === 'push') {
      history.pushState(null, '', url)
    }

    await emit('visit', { url })
    previousUrl = url
  },
}
