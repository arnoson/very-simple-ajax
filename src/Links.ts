import { copyAttributes } from './utils/copyAttributes'
import { copyNode } from './utils/copyNode'
import { emit } from './utils/emit'
import { merge } from './utils/merge'
import { replaceWith } from './utils/replaceWith'

export interface VisitOptions {
  action?: 'push' | 'replace' | 'none'
  useCache?: boolean
  cacheId?: string
}

const parser = new DOMParser()

export class Links {
  private cachedDocuments: Record<string, Document> = {}
  private previousUrl: string | undefined

  start(watchHistory = true) {
    if (watchHistory) {
      window.addEventListener('popstate', () => this.handlePopstate())
    }
  }

  async visit(
    url: string,
    { action = 'push', useCache, cacheId }: VisitOptions = {}
  ) {
    emit('before-visit', { url })

    this.cache(cacheId ?? this.previousUrl ?? window.location.pathname)
    await this.render(url, useCache)

    if (action === 'replace') {
      history.replaceState(null, '', url)
    } else if (action === 'push') {
      history.pushState(null, '', url)
    }

    emit('visit', { url })
    this.previousUrl = url
  }

  private async render(url: string, useCache = false) {
    const newDocument =
      (useCache && this.cachedDocuments[url]) || (await this.load(url))

    if (!newDocument) return

    document.title = newDocument.title
    merge(document.head, newDocument.head)
    const [container, newContainer] = this.getContainers(document, newDocument)

    // Copy each node with `copyNode()`, so `<script>` will get executed.
    const fragment = new DocumentFragment()
    for (const child of Array.from(newContainer.childNodes)) {
      fragment.appendChild(copyNode(child))
    }

    const permanentElements = document.querySelectorAll('[data-permanent]')

    container.innerHTML = ''
    container.appendChild(fragment)
    copyAttributes(container, newContainer)

    permanentElements.forEach((el) => {
      const copy = document.getElementById(el.id)
      copy && replaceWith(copy, el)
    })
  }

  private async load(url: string): Promise<Document | undefined> {
    try {
      const response = await fetch(url)
      return parser.parseFromString(await response.text(), 'text/html')
    } catch (e) {
      // There was a network error. We reload the page so the user sees the
      // browser's network error page.
      window.location.reload()
    }
  }

  private cache(url: string) {
    this.cachedDocuments[url] = document.cloneNode(true) as Document
  }

  private getContainers(document: Document, newDocument: Document) {
    const selector = newDocument.head.querySelector<HTMLMetaElement>(
      "meta[name='very-simple-links:container']"
    )?.content

    const container = selector && document.querySelector(selector)
    const newContainer = selector && newDocument.querySelector(selector)

    if (newContainer && !container)
      console.warn(
        `Container '${selector}' doesn't exist, swapping body instead.`
      )

    return container && newContainer
      ? [container, newContainer]
      : [document.body, newDocument.body]
  }

  private handlePopstate = async () => {
    const url = window.location.pathname
    await this.visit(url, { action: 'none', useCache: true })
  }
}
