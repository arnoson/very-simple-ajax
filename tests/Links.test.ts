import { describe, it, expect, vi } from 'vitest'
import { Links } from '../src/Links'
import fs from 'fs/promises'
import { resolve } from 'path'

const globalRef = global as any

const mockFetch = (content: string) => {
  const fetchResponse = {
    text: () => Promise.resolve(content),
  }
  globalRef.fetch = vi.fn(() => Promise.resolve(fetchResponse))
}

const loadFile = async (file: string) =>
  (await fs.readFile(resolve(__dirname, file))).toString()

describe('Links', () => {
  it('visits a new page', async () => {
    const links = new Links()

    const documentMarkup =
      '<html><head><title>Test</title></head><body>Test</body></html>'

    mockFetch(documentMarkup)
    await links.visit('test.html')

    expect(document.documentElement.outerHTML).toBe(documentMarkup)
  })

  it('merges the heads and swaps the body', async () => {
    const links = new Links()

    const oldPage = await loadFile('old-page.html')
    const newPage = await loadFile('new-page.html')

    document.documentElement.innerHTML = oldPage
    mockFetch(newPage)

    await links.visit('new-page.html')
    expect(document.documentElement).toMatchSnapshot()
  })
})
