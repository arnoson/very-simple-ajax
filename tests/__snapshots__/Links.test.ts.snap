// Vitest Snapshot v1

exports[`Links > merges the heads and swaps the body 1`] = `
<html>
  <head>
    
    
    <meta
      charset="UTF-8"
    />
    
    
    <meta
      content="IE=edge"
      http-equiv="X-UA-Compatible"
    />
    
    
    <meta
      content="width=device-width, initial-scale=1.0"
      name="viewport"
    />
    
    
    <title>
      New Title
    </title>
    
    
    <script
      src="old-script.js"
    />
    
    
    <script
      defer=""
      src="common-script.js"
    />
    
  
    <script
      src="new-script.js"
    />
  </head>
  
  
  <body
    class="new-body"
  >
    
    
    <h1>
      New Body
    </h1>
    
  


  </body>
</html>
`;
